File: sdk/typescript/test-streaming.mjs
/**
 * Test: run a node and get streaming output via WebSocket transport.
 *
 * Usage:
 *   node test-streaming.mjs <api-key> <workflow-id> <node-id>
 *
 * Example:
 *   node test-streaming.mjs nk_live_... 377a8f23-... automation-serverless-function-...
 */

import { init, execution, nodes } from './dist/sdk.esm.js';

const apiKey = process.argv[2];
const workflowId = process.argv[3];
const nodeId = process.argv[4];

if (!apiKey || !workflowId || !nodeId) {
  console.error('Usage: node test-streaming.mjs <api-key> <workflow-id> <node-id>');
  process.exit(1);
}

console.log('Connecting...');
await init({
  transport: 'websocket',
  url: 'http://localhost:8005',
  apiKey,
  workflowId,
});
console.log('✓ Connected\n');

// First read existing output to compare timestamps
console.log('--- Reading existing output ---');
const before = await nodes.getOutput(nodeId);
console.log('  Existing timestamp:', before?.result?.generated_at || 'none');

// Run and stream
console.log('\n--- execution.runNodesAndGetOutput ---');
const stream = execution.runNodesAndGetOutput([nodeId], [nodeId]);

stream.on('output', (nid, data) => {
  console.log(`  [stream:output] nodeId=${nid}`);
  console.log(`    data:`, JSON.stringify(data).substring(0, 200));
});

stream.on('error', (nid, error) => {
  console.log(`  [stream:error] nodeId=${nid} error=${error}`);
});

stream.on('done', () => {
  console.log('  [stream:done] All targets completed');
});

console.log('  Waiting for stream.all()...');
try {
  const results = await stream.all();
  console.log('\n--- Results ---');
  for (const [nid, data] of Object.entries(results)) {
    console.log(`  ${nid}:`, JSON.stringify(data).substring(0, 200));
  }
} catch (e) {
  console.error('  Error:', e.message);
}

// Verify the output is fresh
console.log('\n--- Verifying fresh output ---');
const after = await nodes.getOutput(nodeId);
console.log('  New timestamp:', after?.result?.generated_at || 'none');
console.log('  Is different from before:', (before?.result?.generated_at || '') !== (after?.result?.generated_at || ''));

console.log('\n✓ Streaming test complete');
process.exit(0);

File: sdk/python/test_streaming.py
"""
Test: run a node and get streaming output via Socket.IO transport.

Usage:
    python test_streaming.py <api-key> <workflow-id> <node-id>

Example:
    python test_streaming.py nk_live_... 377a8f23-... automation-serverless-function-...
"""

import asyncio
import json
import sys

sys.path.insert(0, ".")

from noclick import Client


async def main():
    api_key = sys.argv[1] if len(sys.argv) > 1 else None
    workflow_id = sys.argv[2] if len(sys.argv) > 2 else None
    node_id = sys.argv[3] if len(sys.argv) > 3 else None

    if not all([api_key, workflow_id, node_id]):
        print("Usage: python test_streaming.py <api-key> <workflow-id> <node-id>")
        sys.exit(1)

    print("Connecting...")
    sdk = Client(
        api_key=api_key,
        url="http://localhost:8005",  # local dev
        workflow_id=workflow_id,
    )
    await sdk.connect()
    print("✓ Connected\n")

    # Read existing output
    print("--- Reading existing output ---")
    before = await sdk.nodes.get_output(node_id)
    before_ts = None
    if before and isinstance(before, dict):
        before_ts = before.get("result", {}).get("generated_at")
    print(f"  Existing timestamp: {before_ts or 'none'}")

    # Run and get output with streaming
    print("\n--- execution.run_nodes_and_get_output ---")
    print(f"  Running node: {node_id}")
    print("  Waiting for output...")

    try:
        results = await sdk.execution.run_nodes_and_get_output(
            [node_id], [node_id], timeout=30
        )
        print("\n--- Results ---")
        for nid, data in results.items():
            print(f"  {nid}: {json.dumps(data)[:200]}")
    except Exception as e:
        print(f"  Error: {e}")

    # Verify fresh output
    print("\n--- Verifying fresh output ---")
    after = await sdk.nodes.get_output(node_id)
    after_ts = None
    if after and isinstance(after, dict):
        after_ts = after.get("result", {}).get("generated_at")
    print(f"  New timestamp: {after_ts or 'none'}")
    print(f"  Is different from before: {before_ts != after_ts}")

    print("\n✓ Streaming test complete")
    await sdk.disconnect()


if __name__ == "__main__":
    asyncio.run(main())
